### ISSUE TITLE
ets:insert/2 for bag table stores values for same key in reversed order

### ISSUE THREAD

**Issue by madlep** [erlang/otp#6730]
**Describe the bug**
When a list of values are passed to `ets:insert/2` for a `bag` table type, and there are multiple values for the same key, they are incorrectly stored in reversed order. In OTP versions prior to R23, they were stored in the order passed.

We are upgrading from OTP 22 to OTP 25, and have logic that relies on insert order being preserved.

**To Reproduce**
Short code to reproduce:
```erlang
-module(ets_insert_reversed_regression).

-export([example/0, example_multiple_insert/0]).

example() ->
  Tab = ets:new(my_table, [bag]),

  OriginalValues = [{a, 1}, {a, 2}, {a, 3}],
  ets:insert(Tab, OriginalValues),
  ReturnedValues = ets:lookup(Tab, a),

  io:fwrite("Expected =~w~n", [OriginalValues]),
  io:fwrite("Actual   =~w~n", [ReturnedValues]).

example_multiple_insert() ->
  Tab = ets:new(my_table, [bag]),

  OriginalValues1 = [{a, 1}, {a, 2}, {a, 3}],
  ets:insert(Tab, OriginalValues1),

  OriginalValues2 = [{a, 4}, {a, 5}, {a, 6}],
  ets:insert(Tab, OriginalValues2),

  ReturnedValues = ets:lookup(Tab, a),

  io:fwrite("Expected =~w~n", [OriginalValues1 ++ OriginalValues2]),
  io:fwrite("Actual   =~w~n", [ReturnedValues]).
```

**Correct** original behaviour running on OTP 22 - note ordering of values for key `a` is preserved:
```
Erlang/OTP 22 [erts-10.7.2.18] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]

Eshell V10.7.2.18  (abort with ^G)
1> c(ets_insert_reversed_regression).
{ok,ets_insert_reversed_regression}
2> ets_insert_reversed_regression:example().
Expected =[{a,1},{a,2},{a,3}]
Actual   =[{a,1},{a,2},{a,3}]
ok
3> ets_insert_reversed_regression:example_multiple_insert().
Expected =[{a,1},{a,2},{a,3},{a,4},{a,5},{a,6}]
Actual   =[{a,1},{a,2},{a,3},{a,4},{a,5},{a,6}]
ok
```

**Incorrect** regression behaviour running on latest OTP 25 - note order of insert calls themselves is preserved, but values for same key in each call are reversed (this happens on releases after OTP 23):
```
Erlang/OTP 25 [erts-13.1.4] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [jit]

Eshell V13.1.4  (abort with ^G)
1> c(ets_insert_reversed_regression).
{ok,ets_insert_reversed_regression}
2> ets_insert_reversed_regression:example().
Expected =[{a,1},{a,2},{a,3}]
Actual   =[{a,3},{a,2},{a,1}]
ok
3> ets_insert_reversed_regression:example_multiple_insert().
Expected =[{a,1},{a,2},{a,3},{a,4},{a,5},{a,6}]
Actual   =[{a,3},{a,2},{a,1},{a,6},{a,5},{a,4}]
ok
```

**Expected behavior**
Behaviour of OTP 22 is reinstated. The docs for [`ets:lookup/2`](https://www.erlang.org/doc/man/ets.html#lookup-2) _seem_ to imply this is the correct behaviour (but they're not specific on whether "time order" counts for values in the same `insert` call)

> Notice that the time order of object insertions is preserved; the first object inserted with the specified key is the first in the resulting list, and so on.

**Affected versions**
OTP 23 and up.

**Additional context**
Reading through the source code, it looks like this was introduced as part of https://github.com/erlang/otp/pull/2396 .

In particular, this change looks suspicious: https://github.com/erlang/otp/blob/d5d89208521cc0d025ed47c6ea8419ec7aadea48/erts/emulator/beam/erl_db.c#L1574 

It appears to be converting the list passed to `ets:insert/2` from `Eterm` to `DbTerm` values, and building a linked list of values for a key by preprending to a linked list - but then leaving it in reversed order as a result of prepending, and storing that in the table for the key.

It also looks like this will affect `duplicate_bag` as well with large enough inserts, but it needs to exceed the reduction count for this logic to kick in, and I haven't tested that situation yet.

If multiple individual calls to `ets:insert/2` are made using a single tuple instead of a list of tuples, this behaviour can be worked around, _however_, if the tab is saved with `ets:tab2file/2/3` and then reloaded with `ets:file2tab/1/2` it goes through the code path affected by this bug, and we get the same issue.

**Comment by madlep** [comment]
I'm happy to attempt a fix for this. I'm not hugely familiar with the emulator C code though.

It seems that either reversing the linked list of `DbTerm` values, or building it in append order (rather than prepend) would work, but I'm not sure of what flow on effects that would have.

**Comment by sverker** [comment]
Thanks for the excellent report.
If you want to try fix it, I think the best approach is to change it to db_dbterm_list_**append**() and add to the tail of the list.
Note that all code inside `#ifdef YCF_FUNCTIONS` is passed through the YCF code generator that injects yield point. Such C code have some [restrictions](https://github.com/erlang/otp/blob/master/erts/lib_src/yielding_c_fun/README.md#code-restrictions), like not allowing pointers to local variables. But I think something like this would work:
```
#ifdef YCF_FUNCTIONS
static void* ets_insert_2_list_copy_term_list(...)
{
    void* db_term_list = NULL;
    void* list_tail = NULL;
    :
      if (db_term_list != NULL) {
          meth->db_dbterm_list_append(list_tail, term);
      } else {
          db_term_list = term;
      }
      list_tail = term;
    }

    return db_term_list;
```

If you want to give it a try, base fix on OTP-23.3.4

**Comment by gomoripeti** [comment]
interesting finding.

Sorry to piggy-back this issue but the documentation also states
> Insert and lookup times in tables of type set, bag, and duplicate_bag are constant, regardless of the table size.

I think this is also not correct as inserting multiple values for the same key in a bag table, the insert time is growing with the number of existing values for the same key. (Or does it depend on prepending or appending?)

**Comment by sverker** [comment]
> I think this is also not correct as inserting multiple values for the same key in a bag table, the insert time is growing with the number of existing values for the same key.

Yes, `set`, `bag` and `duplicate_bag` are hash tables where we only hash on the key term. That is, each hash bucket contains a linked list of all tuples with the same key and all other keys that happen to collide hash wise on that bucket.
For `set` this is not a problem as there is only one tuple per key and hash collisions are reasonable rare. The table grows dynamically.
For `bag` and `duplicate_bag` we may get a linear search to skip past other keys and for insert in `bag` make sure we do no insert a duplicate. An unsuccessful lookup will for example have to step through the entire list.
I think a warning would even be appropriate to not insert extensive amount of tuples with the same key. It not only leads to linear searches, these linear searches through a bucket are not yielding. It may thus hurt soft real time characteristics.

**Comment by max-au** [comment]
Minor note, if you change the behaviour *now*, you may break those who have relied on this behaviour in OTP 23, 24 and 25...

**Comment by sverker** [comment]
> Minor note, if you change the behaviour _now_, you may break those who have relied on this behaviour in OTP 23, 24 and 25...

Sure, but it must be harder to argue that "_time order of object insertion_" in
> Notice that the time order of object insertions is preserved; the first object inserted with the specified key is the first in the resulting list, and so on.

actually means the reverse order of the list.

**Comment by sverker** [comment]
@madlep What are your plans? If you don't, we will make a fix for 25.3 which has code freeze 22 February.

**Comment by sverker** [comment]
We probably would want the fix in OTP 26.0-rc1 which has code freeze 10 February.

**Comment by madlep** [comment]
@sverker Having a look at this over the weekend. Planning to get something PR ready tomorrow.

**Comment by sverker** [comment]
#6752 Merged to maint (25.3) and master (26.0).
The plan is to leave it at that and not release fixes on 23 and 24 due to the risk of causing semantic problem on deployed systems.
